packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -214,6 +215,15 @@ export function makeVirtualObjectManager(
     }
   }
 
+  function isExternallyHeld(value) {
```

Comment:
But I'm confused by `isExternallyHeld`. Given how it's used, it looks like the change it introduces is:
* if a Presence or Representative is used as a key, put it in a Set so it's held by a strong reference forever

Whereas what I think we need (for various pieces of the GC extravaganza) is:
* if a Presence is used as a key, put its *vref* in a Set to remember that it's recognizable (but not necessarily reachable), to inhibit a `syscall.retireImport` on that vref
  * liveslots will start out in phase 1 by just never doing `retireImport`, so this is about enabling phase 2, where we retire (the majority of) Presences that *weren't* used as a key, and then down the road we do phase 3 where we retire the remaining Presences that were once used as a key but then got removed
* if a Remotable is used as a key, ignore its vref and use the object itself, so the WeakMap entry will be deleted if/when the Remotable goes away
* if a Representative is used as a key, we don't currently need to do anything special: we conservatively assume that all virtual object vrefs are recognizable (until we get to phase 3 and do much more detailed refcounting to let us automatically delete WeakMap entries whose keys have become unreachable), and simply being a key in a WeakMap doesn't force it to be reachable (vs being a value, or a key of a (strong) Map)

There's a separate set of needs for things used as *values*, but that's covered in #3132 and #3133. For keys, I think the changes you made to `virtualObjectKey` and `vrefKey` are sufficient.

If that makes sense, I think the necessary change is to remove `isExternallyHeld` and the code that calls it. If it doesn't make sense, let's re-sync our understanding of the requirements.. it's possible I've gotten myself turned around.

Reply:
It's possible you have gotten yourself turned around, but maybe me too.  I'm trying to reconstruct my brain state from when I wrote that code, as I believed I was implementing the design that we worked out.  However, let me try to walk through it again myself to see what makes sense:

In the case of a Representative: the Representative is an object in memory, by definition.  When it is in memory, there is a `valToSlot` entry that it indexes, but since `valToSlot` is a `WeakMap` that will go away when the last live reference to the in memory object goes away, but then it doesn't matter since, tautologically, nobody can ever refer to the live in memory object at that point.  The only way we get back to having a Representative again is if we try to deserialize its vref, which will be recognized as a virtual object reference and cause a new Representative to be summoned into existence, but since there wasn't a Representative there before it will be come the new sole Representative and nobody will be the wiser.  Another thing that might have referred to it is a `WeakMap` key, but that was already taken care of by indexing it in a side table by its vref, and since it's only being recognized and not read, nothing will ever try to go from the `WeakMap` key to the object itself.  So there's no reason to try to pin the Representative in memory.

In the case of a Presence: the Presence is also an object in memory that is held there by, ultimately, the `slotToVal` table entry associated with its importation.  If that import goes away, the Presence object could go away too, but if somebody had used it as a `WeakMap` key we have to be able to get back to that entry in case somebody from off vat refers to it again later, which means we have to retain its clist entry so that it doesn't get assigned a new vref if it is reimported, even if we don't actually care about the Presence object itself (i.e., your Reachable vs. Recognizable distinction).  But it's the clist entry that keeps that information; we don't need to pin the Presence.

So if I'm working this through correctly, in neither the Representative nor the Presence case does the `VirtualObjectAwareWeakMap` need to pin the in-memory object, and so the `virtuallyReferencedObjects` set and the `isExternallyHeld` predicate serve no purpose but to keep things in RAM that don't need to be there.

Reply:
On the basis of the above analysis, which aligns with yours, I'm going to take those things out and push a new commit.  Confirm or reject at your next opportunity.

Reply:
I concur with everything except a minor pedantic quibble:

> the Presence is also an object in memory that is held there by, ultimately, the slotToVal table entry associated with its importation

The Presence is ultimately held in memory by userspace code (initially the parameter in the method call that received it, or the first argument of a `.then` callback for a promise resolution). The `slotToVal` entry is a WeakRef, which can cause a new delivery that cites the same vref to get deserialized into the same Presence as is already in memory, but `slotToVal` cannot itself keep that Presence alive. Once userspace gives up the Presence, the WeakRef eventually goes dead, a finalizer gets called, and (depending upon whether the vref is still referenced by some offline data somewhere) we might be able to inform the kernel that this vat can no longer reach the vref.

Awesome, thanks, I'll look for the new commit.

Reply:
Er, yeah.  That was a, call it a typo?, I wrote `valToSlot` when I meant `slotToVal`, i.e., even if nobody in the vat is referring to it the importation can still refer to it until it's retired.  But, yeah, we're getting pedantic at this point.